Add windows docker build script

youngbash88 opened the pull request:
### Summary of the PR

- Added `docker.ps1` to enable building the container using PowerShell with the following command:
  ```powershell
  .\docker.ps1 build
  ```
  And publishing the container using:
  ```powershell
  .\docker.ps1 publish
  ```
 as in `docker.sh` to resolve https://github.com/rust-vmm/rust-vmm-container/issues/49 (Note: didn't update the manifest in `docker.sh`)

- Modified `.github/workflows/docker-publish.yml` to: Add `build-windows` job for Windows containers.
### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [x] All commits in this PR have Signed-Off-By trailers (with
  `git commit -s`), and the commit message has max 60 characters for the
  summary and max 75 characters for each description line.
- [ ] All added/changed functionality has a corresponding unit/integration
  test.
- [ ] All added/changed public-facing functionality has entries in the "Upcoming 
  Release" section of CHANGELOG.md (if no such section exists, please create one).
- [ ] Any newly added `unsafe` code is properly documented.

stefano-garzarella commented on docker.ps1:
```diff
@@ -0,0 +1,63 @@
+ [string]$Command
+)
+$ErrorActionPreference = "Stop"
+$GIT_COMMIT = (git rev-parse HEAD)
+$GIT_BRANCH = (git rev-parse --abbrev-ref HEAD)
+$IMAGE_NAME = "rustvmm/dev"
+$REGISTRY = "index.docker.io"
```
I have no experience with power shell, but is there a way to have a common file (e.g. `docker.env`) between `docker.sh` and `docker.ps1` where we define common variables or commands?

I mean something that in bash we can include with `source` where we define variables that we use in both, so we have one place to edit them.

I'm not sure if it's easy to do though, I found something here: 
https://stackoverflow.com/questions/72236557/how-do-i-read-a-env-file-from-a-ps1-script

But again, I have no experience, so it's up to you whether it makes sense or not.

stefano-garzarella reviewed (commented):
Please, can you update `README.md` too?

stefano-garzarella commented:
@youngbash88 thanks for the PR, I just enabled the CI, but it's failing soon, can you take a look?

roypat commented:
Hey, I'm not really familiar with the context behind #49, could you provide a bit more motivation for this? Will it help with testing some of the mshv stuff in GHA? Because our buildkite CI infra doesnt have any windows hosts, so a windows docker container can't run anywhere in CI that I'm aware of.

youngbash88 commented:
> Hey, I'm not really familiar with the context behind #49, could you provide a bit more motivation for this? 

Hey, thanks for the clarification! The motivation behind #49 was to add a script similar to docker.sh for Windows. I initially modified the CI to add build-windows, thinking it could be useful for future testing.

> Because our buildkite CI infra doesnt have any windows hosts, so a windows docker container can't run anywhere in CI that I'm aware of.

Since the Buildkite CI infra doesn’t have Windows hosts, the Windows build won’t work in the current setup. I’ll go ahead and remove the Windows build step from the CI configuration. Should I open an issue to track this for potential Windows host support in the future?

youngbash88 commented on README.md:
````diff
@@ -123,4 +160,4 @@ If it is the first time you are creating a docker manifest, most likely it will
 are enabled```. Checkout
 [this article](https://medium.com/@mauridb/docker-multi-architecture-images-365a44c26be6)
-to understand why and how to fix it.
+to understand why and how to fix it.
````
I dont know why does show as if there is a change here

stefano-garzarella replied:
Could this be related to `core.autocrlf` git option? Can you try setting it to false?

youngbash88 replied:
I tried that but it's still the same

stefano-garzarella commented:
@youngbash88 there is something wrong with the rebase here, can you do it properly?

youngbash88 commented:
@stefano-garzarella done.

stefano-garzarella commented:
@youngbash88 there is still a commit with just "test"

youngbash88 commented:
@stefano-garzarella I squashed the commits and updated the commit message. LMK if it's good or if any changes are needed

stefano-garzarella commented on docker.sh:
```diff
@@ -1,10 +1,12 @@
-GIT_BRANCH=$(git rev-parse --abbrev-ref HEAD)
-IMAGE_NAME=rustvmm/dev
-REGISTRY=index.docker.io
+# moved to docker.env
+#GIT_COMMIT=$(git rev-parse HEAD)
+#GIT_BRANCH=$(git rev-parse --abbrev-ref HEAD)
+#IMAGE_NAME=rustvmm/dev
+#REGISTRY=index.docker.io
```
Can we just remove them?

We can leave a comment, but on top of `source ...`, something like this:
```
# env vars in common with docker.ps1 are stored in docker.env
source ...
```

stefano-garzarella commented on Dockerfile.windows.x86_64:
```diff
@@ -7,7 +7,10 @@ ARG RUST_TOOLCHAIN
 ENV chocolateyUseWindowsCompression "true"
-ENV RUST_TOOLCHAIN="1.46.0"
+# Update Rust version to one that supports cargo resolver
+ENV RUST_TOOLCHAIN="1.75.0"
```
Should we be aligned with the one defined in `build_container.sh` ?

youngbash88 replied:
I'll add this to `docker.env`

youngbash88 replied:
>  I'll add this to `docker.env`

@stefano-garzarella Just to confirm, does that sound good to you?

stefano-garzarella replied:
Yeah, I think we should try to uniforn it, IIUC for Linux it's defined in `build_container.sh`, so maybe we should remove from it and pass it from the Dockerfile

stefano-garzarella commented on README.md:
````diff
@@ -42,6 +42,14 @@ Example of running cargo build on the kvm crate:
     Finished `release` profile [optimized] target(s) in 6.34s
 ```
 
+For Windows users (ensure Docker Desktop is in Linux containers mode):
````
Obviously I don't know what it means, but do we have any steps to suggest checking this?

(If it's an obvious one, sorry for the question).

youngbash88 replied:
If someone is familiar with Docker on Windows, it's quite straightforward. But I can add instructions or a link for checking and switching to linux mode

stefano-garzarella replied:
If it's straightforward, don't worry!

stefano-garzarella commented on README.md:
````diff
@@ -42,6 +42,14 @@ Example of running cargo build on the kvm crate:
 
+For Windows users (ensure Docker Desktop is in Linux containers mode):
+```powershell
+> git clone https://github.com/rust-vmm/kvm.git
+> cd kvm
+> docker run --volume "${PWD}:/kvm" `
+    rustvmm/dev:latest `
+    /bin/bash -c "cd /kvm && cargo build --release"
````
What are the differences with the steps for linux above?

youngbash88 replied:
In powershell ( \ ) needs to be replaced with ( ` )

stefano-garzarella replied:
ooo, I see, thanks :-)

stefano-garzarella reviewed (commented):
I left minor comments, but I didn't test the windows script since I don't have access to windows machine. The changes to linux scripts LGTM!

RuoqingHe commented:
I think we can ask people in `mshv` to get involved on this topic, IMO they most likely need this

stefano-garzarella commented:
> I think we can ask people in `mshv` to get involved on this topic, IMO they most likely need this

Good point, let me ping the code owners of that crate.

@russell-islam @liuw @NunoDasNeves @jinankjain @praveen-pk please, can you take a look?

liuw commented:
> > I think we can ask people in `mshv` to get involved on this topic, IMO they most likely need this
> 
> Good point, let me ping the code owners of that crate.
> 
> @russell-islam @liuw @NunoDasNeves @jinankjain @praveen-pk please, can you take a look?

We use Linux for our infrastructure through and through. I myself have close to zero knowledge on PowerShell. :-)

russell-islam commented:
> > > I think we can ask people in `mshv` to get involved on this topic, IMO they most likely need this
> > 
> > 
> > Good point, let me ping the code owners of that crate.
> > @russell-islam @liuw @NunoDasNeves @jinankjain @praveen-pk please, can you take a look?
> 
> We use Linux for our infrastructure through and through. I myself have close to zero knowledge on PowerShell. :-)

+1

stefano-garzarella commented:
@youngbash88 LGTM, but please organize the commits better:
- avoid doing multiple stuff in a single commit (e.g. last commit can be splitted at least in 2 commits)
- avoid changing code introduced by the PR in next commits (e.g. in this case, introduce the docker.env and changes related to linux first, then introduce the docker.ps1, so you can use that file directly
- avoid merging commit from main, just rebase this PR on top of it

youngbash88 commented:
> @youngbash88 LGTM, but please organize the commits better:
> 
>     * avoid doing multiple stuff in a single commit (e.g. last commit can be splitted at least in 2 commits)
> 
>     * avoid changing code introduced by the PR in next commits (e.g. in this case, introduce the docker.env and changes related to linux first, then introduce the docker.ps1, so you can use that file directly
> 
>     * avoid merging commit from main, just rebase this PR on top of it

Ok got it, thanks for the feedback

youngbash88 commented:
@stefano-garzarella I just pushed the last commits can you check them?

russell-islam commented on README.md:
```diff
@@ -107,6 +137,18 @@ On an `aarch64` platform:
+> .\docker.ps1 publish
+
+
+#To check if Docker is in Linux containers mode:
```
Add an space after #

russell-islam commented on docker.ps1:
```diff
@@ -0,0 +1,97 @@
+    $imageExists = docker image inspect $tag 2>$null
+    if ($LASTEXITCODE -ne 0) {
+        Write-Error "Image $tag not found locally. Please run 'build' first."
+        exit 1
```
Why not exiting with $LASTEXITCODE,

russell-islam commented on docker.ps1:
```diff
@@ -0,0 +1,97 @@
+    docker push $tag
+    if ($LASTEXITCODE -ne 0) {
+        Write-Error "Failed to publish $tag"
+        exit $LASTEXITCODE
```
Same as above. I would write a simple funtion

funtion check_exit_code() {
     if ($1 -ne 0) {
          Write-Error $2
          exit $1
      }
 }
 
 Then call check_exit_code code "msg"
 
 avoid code duplicacy
 
 Just a suggestion.

youngbash88 replied:
OK, thanks for the review. I'll follow your suggestions

stefano-garzarella commented:
> @stefano-garzarella I just pushed the last commits can you check them?

I don't see my suggestion fixed:
- commit 24edb1da29cdd637655ea0e6f6255575025485a3 must be split in several commits, the description itself should tell you in how many commit you should split.
- docker.env e docker.sh changes should be introduced first, so you don't need to change docker.ps1 later, just use docker.env since you introduce it
- commit d6866ecba68eceeff01b30e0318e59ba77c3b885 the tile is too long, please use short title and add a better description explaining the "why" of the change (also for other commits)
- commit d3bc232eb9b1ae0b86991fd8e46ae3d9fd1c552f is fine, but mixed with other changes (e.g. removing the old variable commented)
- commit eb670638f9e55c1501f02ae88bb0dffbccc3f916 can be merged with commit d3bc232eb9b1ae0b86991fd8e46ae3d9fd1c552f
- commit e5589c0cd0dfdca09061c3ab1b06b8d2ebf4e17b can also be merged with the commit where you update the readme (that need to be factored out from commit 24edb1da29cdd637655ea0e6f6255575025485a3)

youngbash88 commented:
> I don't see my suggestion fixed:

Oh sorry! I think I miss understood you earlier. I'll go ahead and update the PR.

stefano-garzarella commented:
> > I don't see my suggestion fixed:
> 
> Oh sorry! I think I miss understood you earlier. I'll go ahead and update the PR.

Don't worry ;-) if something is not clear, don't mind asking for clarification ;-) we are not native speakers, so sometimes we are not clear.

youngbash88 commented:
@stefano-garzarella  I've updated the commits. can you take a look?

russell-islam commented on docker.ps1:
```diff
@@ -0,0 +1,99 @@
+        Write-Error $Message
+        exit $ExitCode
+    }
+}
```
Add a new line after this. :)

youngbash88 replied:
Done.

stefano-garzarella commented on docker.sh:
```diff
@@ -1,10 +1,8 @@
 #!/usr/bin/env bash
 set -e
+# Environment variables in common with the Docker.ps1 comes from the docker.env
```
nit: since you need to change something, I'd move this comment in the first commit, or just remove it

stefano-garzarella reviewed (changes requested):
IIUC it looks like there is some issues with CI and RUST_TOOLCHAIN changes.
@youngbash88 can you take a look?

stefano-garzarella commented:
@youngbash88 please check patch 3 description, try to keep lines around 75 columns.

It looks like CI is still failing.

youngbash88 commented:
> @youngbash88 please check patch 3 description, try to keep lines around 75 columns.

@stefano-garzarella I updated [this](https://github.com/rust-vmm/rust-vmm-container/pull/125/commits/f02c683fa14dbac0b444d6d86300b400e5f451a1) commit message, is that what you meant? 
Also i changed the approach, now ` build_container.sh ` source rust toolchain  from ` docker.env ` directly

stefano-garzarella commented on build_container.sh:
```diff
@@ -1,8 +1,8 @@
 #!/usr/bin/env bash
 set -ex
+eval "$(grep '^RUST_TOOLCHAIN=' "$(dirname "$0")/docker.env")"
```
BTW I think to fix the issue you had, you can add something like this:

```diff
diff --git a/.github/workflows/docker-publish.yml b/.github/workflows/docker-publish.yml
index de3f2b7..8314ab2 100644
--- a/.github/workflows/docker-publish.yml
+++ b/.github/workflows/docker-publish.yml
@@ -57,6 +57,10 @@ jobs:
           echo "IMAGE_NAME=${IMAGE_NAME}" >> $GITHUB_ENV
           echo "Image name to be published is: ${IMAGE_NAME}"
 
+          RUST_TOOLCHAIN=$(./docker.sh print-rust-toolchain)
+          echo "RUST_TOOLCHAIN=${RUST_TOOLCHAIN}" >> $GITHUB_ENV
+          echo "Rust toolchain used is: ${RUST_TOOLCHAIN}"
+
           NEXT_VERSION=$(./docker.sh print-next-version)
           echo "VERSION=${NEXT_VERSION}" >> $GITHUB_ENV
           echo "Next version to be published is: ${NEXT_VERSION}"
@@ -72,6 +76,7 @@ jobs:
           # docker container on both x86_64 and arm64.
           platforms: linux/amd64,linux/arm64
           tags: ${{ env.VERSION }},${{ env.IMAGE_NAME }}:latest
+          build-args: RUST_TOOLCHAIN==${{ env.RUST_TOOLCHAIN }}
           cache-from: type=gha
           cache-to: type=gha,mode=max
 
@@ -116,6 +121,10 @@ jobs:
           echo "IMAGE_NAME=${IMAGE_NAME}" >> $GITHUB_ENV
           echo "Image name to be published is: ${IMAGE_NAME}"
 
+          RUST_TOOLCHAIN=$(./docker.sh print-rust-toolchain)
+          echo "RUST_TOOLCHAIN=${RUST_TOOLCHAIN}" >> $GITHUB_ENV
+          echo "Rust toolchain used is: ${RUST_TOOLCHAIN=}"
+
           NEXT_VERSION=$(./docker.sh print-next-version)
           echo "VERSION=${NEXT_VERSION}" >> $GITHUB_ENV
           echo "Next version to be published is: ${NEXT_VERSION}"
@@ -129,6 +138,7 @@ jobs:
           push: ${{ github.event_name != 'pull_request' }}
           platforms: linux/amd64
           tags: ${{ env.VERSION }}-riscv,${{ env.IMAGE_NAME }}:latest-riscv
+          build-args: RUST_TOOLCHAIN==${{ env.RUST_TOOLCHAIN }}
           cache-from: type=gha
           cache-to: type=gha,mode=max
```

Of course implementing it in docker.sh

youngbash88 replied:
> can we just `source` it?

I tried but it was failing to build locally. I'll try again

youngbash88 replied:
> BTW I think to fix the issue you had, you can add something like this:
> 
> ```diff
> diff --git a/.github/workflows/docker-publish.yml b/.github/workflows/docker-publish.yml
> index de3f2b7..8314ab2 100644
> --- a/.github/workflows/docker-publish.yml
> +++ b/.github/workflows/docker-publish.yml
> @@ -57,6 +57,10 @@ jobs:
>            echo "IMAGE_NAME=${IMAGE_NAME}" >> $GITHUB_ENV
>            echo "Image name to be published is: ${IMAGE_NAME}"
>  
> +          RUST_TOOLCHAIN=$(./docker.sh print-rust-toolchain)
> +          echo "RUST_TOOLCHAIN=${RUST_TOOLCHAIN}" >> $GITHUB_ENV
> +          echo "Rust toolchain used is: ${RUST_TOOLCHAIN}"
> +
>            NEXT_VERSION=$(./docker.sh print-next-version)
>            echo "VERSION=${NEXT_VERSION}" >> $GITHUB_ENV
>            echo "Next version to be published is: ${NEXT_VERSION}"
> @@ -72,6 +76,7 @@ jobs:
>            # docker container on both x86_64 and arm64.
>            platforms: linux/amd64,linux/arm64
>            tags: ${{ env.VERSION }},${{ env.IMAGE_NAME }}:latest
> +          build-args: RUST_TOOLCHAIN==${{ env.RUST_TOOLCHAIN }}
>            cache-from: type=gha
>            cache-to: type=gha,mode=max
>  
> @@ -116,6 +121,10 @@ jobs:
>            echo "IMAGE_NAME=${IMAGE_NAME}" >> $GITHUB_ENV
>            echo "Image name to be published is: ${IMAGE_NAME}"
>  
> +          RUST_TOOLCHAIN=$(./docker.sh print-rust-toolchain)
> +          echo "RUST_TOOLCHAIN=${RUST_TOOLCHAIN}" >> $GITHUB_ENV
> +          echo "Rust toolchain used is: ${RUST_TOOLCHAIN=}"
> +
>            NEXT_VERSION=$(./docker.sh print-next-version)
>            echo "VERSION=${NEXT_VERSION}" >> $GITHUB_ENV
>            echo "Next version to be published is: ${NEXT_VERSION}"
> @@ -129,6 +138,7 @@ jobs:
>            push: ${{ github.event_name != 'pull_request' }}
>            platforms: linux/amd64
>            tags: ${{ env.VERSION }}-riscv,${{ env.IMAGE_NAME }}:latest-riscv
> +          build-args: RUST_TOOLCHAIN==${{ env.RUST_TOOLCHAIN }}
>            cache-from: type=gha
>            cache-to: type=gha,mode=max
> ```
> 
> Of course implementing it in docker.sh


Ok, I'll try this

stefano-garzarella replied:
> Ok, I'll try this

Yeah, of course you need to bring back the `--build-arg ...` in docker.sh too.

Now I'm wondering if GIT_BRANCH/GIT_COMMIT are really used, maybe it's a leftover, but we can fix later.
In the end I think every `--build-arg` you add in `docker.sh` should be added in `docker-publish.yml` too.

youngbash88 replied:
The reason i changed the is because i went thru the commits to understand why it was in ` build_contianer.sh ` in the first place and found [this](https://github.com/youngbash88/rust-vmm-container/commit/4a785a448108fc0d2ccf8da368213d4efcb1067a) which made me think there is something wrong with ` --build-arg`  for rust toolchain.


> Now I'm wondering if GIT_BRANCH/GIT_COMMIT are really used, maybe it's a leftover 

this one i think is used in ` print-next-version ` but i don't know if they do something else

youngbash88 replied:
I can do .yml file changes in [this](https://github.com/rust-vmm/rust-vmm-container/pull/125/commits/7aee8aad1f0bc64bb00acb5408db12cabfbf4603) commit or should it be separate ?

stefano-garzarella replied:
> The reason i changed the is because i went thru the commits to understand why it was in build_contianer.sh in the first place and found [this](https://github.com/youngbash88/rust-vmm-container/commit/4a785a448108fc0d2ccf8da368213d4efcb1067a) which made me think there is something wrong with  --build-arg for rust toolchain.

I see, feel free to avoid that change and duplicate it on both dockers file if it's too much of work.

> this one i think is used in `print-next-version` but i don't know if they do something else

I meant in the Dockerfile, IIUC we don't need to pass them with --build-arg, but as I mentioned we can figure out later.

stefano-garzarella replied:
> I can do .yml file changes in [this](https://github.com/rust-vmm/rust-vmm-container/pull/125/commits/7aee8aad1f0bc64bb00acb5408db12cabfbf4603) commit or should it be separate ?

What kind of change?

stefano-garzarella replied:
If we do that, I don't think we need the commit you linked anymore, no?

youngbash88 replied:
You're right. I'll go ahead and do the changes

stefano-garzarella reviewed (commented):
Please try to write better commit. I suggest taking a look at https://docs.kernel.org/process/submitting-patches.html#describe-your-changes

For example `Modify docker-publish.yml ` is not a good title, better something like:
```
Add docker.env to workflow paths

build_container.sh sources RUST_TOOLCHAIN from it, so we need to ...
```

So summarize in the title what the patch is doing, then you can add details in the body of the commit message, which provides more context and also the reason (like you did).

That said, I think that commit can be removed if we implement what I suggested, like we do for other env vars.

youngbash88 commented:
@stefano-garzarella can you please enable the CI?

stefano-garzarella commented on Dockerfile.windows.x86_64:
```diff
@@ -7,7 +7,8 @@ ARG RUST_TOOLCHAIN
 FROM mcr.microsoft.com/windows/servercore:$WIN_VER
 
 ENV chocolateyUseWindowsCompression "true"
-ENV RUST_TOOLCHAIN="1.46.0"
```
Should we do like we did in `Dockerfile`?

I mean:
```
ARG RUST_TOOLCHAIN
ENV RUST_TOOLCHAIN=${RUST_TOOLCHAIN}
```

If yes, please do in the previous commit.

youngbash88 replied:
I did build the image multiple time the only issue i had was related to chocolaty but there wasn't a problem with rust toolchain.

stefano-garzarella replied:
okay, so why we need `ENV RUST_TOOLCHAIN=${RUST_TOOLCHAIN}` in Dockerfile?

In any case, I'd do this change in the previous commit.

stefano-garzarella replied:
Ah I see, because in Dockerfile we use it in a nested script, while in Dockerfile.windows.x86_64 not.

Okay, I'm fine but please move this change to the previous commit and update the message in this commit.

youngbash88 replied:
Ok, now the ` build-riscv`  is failing and I can't tell why

stefano-garzarella replied:
https://github.com/rust-vmm/rust-vmm-container/actions/runs/14225467334/job/39865283023?pr=125#step:6:13539

```
 913.8 + sh -s -- -y --default-toolchain '' --profile minimal --component clippy,rustfmt
915.5 info: downloading installer
919.9 error: error: invalid value '' for '--default-toolchain <DEFAULT_TOOLCHAIN>': invalid toolchain name ''
```

It seems related, I guess we need 
```
ARG RUST_TOOLCHAIN
ENV RUST_TOOLCHAIN=${RUST_TOOLCHAIN}
```

also in Dockerfile.riscv64

stefano-garzarella commented:
I don't want to be picky, but please check commit messages. Use imperative (not past verb), also explain reason, so in the future we can remember what we did.

Patch 1:
```
Add docker.env

We are going to add a windows script, so concentrate the environment
variables in one file to use them in the linux and windows script.
```

Patch 2:
```
Define RUST_TOOLCHAIN in docker.env

Instead of defining RUST_TOOLCHAIN in both Dockerfile and
Dockerfile.windows.x86_64files, let's define it in one place
(docker.env and pass it as an environment variable on both
docker.sh and github workflows.
```

etc.

stefano-garzarella commented:
@youngbash88 thanks, LGTM!

youngbash88 commented:
@stefano-garzarella Thanks for your time!

RuoqingHe reviewed (commented):
Great work - could @youngbash88 provide some screenshots showing this actually works for testing rust-vmm crates on windows, and some OS details as well, that would be very helpful 🙂

youngbash88 commented:
@RuoqingHe 
<img width="960" alt="Screenshot 2025-04-03 143654" src="https://github.com/user-attachments/assets/cc53f8cc-ccb2-4897-809f-42fff863bfda" />
<img width="960" alt="Screenshot 2025-04-03 144141" src="https://github.com/user-attachments/assets/aa1b36c0-5b36-4f9a-84a7-1356b0677c2a" />
<img width="960" alt="Screenshot 2025-04-03 144101" src="https://github.com/user-attachments/assets/0436a5e3-492d-4e58-8002-3455750c5f4d" />
<img width="960" alt="Screenshot 2025-04-03 144757" src="https://github.com/user-attachments/assets/3373f065-193a-4152-bb75-6db53b53e8fa" />
<img width="242" alt="Screenshot 2025-04-03 145315" src="https://github.com/user-attachments/assets/e47383af-8103-4b1d-967b-cc0208896f60" />

RuoqingHe reviewed (approved):
Looks good to me, thanks @youngbash88 ❤️
